proposals/interface-types/Explainer.md

```diff
@@ -1,826 +1,1560 @@
+            | char
+            | (list <intertype>)
+            | (record (field <name> <id>? <intertype>)*)
+            | (variant (case <name> <id>? <intertype>?)*)
+```
+`f32` and `f64` are the same types that appear in `valtype` and thus `valtype`
+and `intertype` intersect at these two types. `char` is defined to be a
+[Unicode scalar value]  (i.e., a non-[surrogate]  [code point]).
+
+Unlike core wasm's `valtype`, which, starting with [function references],
+allows cyclic type definitions, `intertype` type definitions are required to be
+acyclic. This restriction follows from the fact that interface types are meant
+to be used at inter-language boundaries to copy values and matches the
+restrictions of other IPC/RPC schema languages.
+
+Additionally, while core wasm's `valtype` contains reference values which refer
+to mutable state, `intertype` compound values are transitively immutable. One
+important consequence of this distinction is that, while `valtype` reference
+subtyping is non-coerceive and constrained by low-level memory layout
+considerations, `intertype` subtyping can be highly coerceive, thereby allowing
+flexible, backwards-compatible evolution of interface-typed APIs. Since copying
+is inherent in the use of interface types, coercions can be folded into the
+copy. A summary of the allowed coercions is given by the following table:
+
+| Type             | Coercion allowed |
+| ---------------- | ---------------- |
+| `f32`, `f64`     | `f32` to `f64`   |
+| `s8`, `u8`, `s16`, `u16`, `s32`, `u32`, `s64`, `u64` | if the source range is included in destination range |
+| `char`           | none |
+| `list`           | if the list's element type can be coerced |
+| `record`         | if the record's fields' types can be coerced, matching fields by name regardless of order and allowing source fields to be ignored |
+| `variant`        | if the variant's cases' types can be coerced, matching cases by name regardless of order and allowing destination cases to be ignored |
+
+While these types are sufficiently general to capture a number of
+more-specialized types that often arise in interfaces, it can still be
+beneficial to have the specialized types explicitly represented in API
+signatures. For example, this allows specialized source-language types to be
+used in automatically-generated source-level API declarations. Thus, the
+proposal contains several text-format and binary-format [abbreviations]. In the
+notation below, the types on the left-hand side of the `≡` are expanded during
+parsing/decoding into the abstract syntax of the right-hand side. Thus,
+validation and execution are purely defined in terms of the general type; the
+abbreviations only exist in the concrete formats.
+```
+                                      string ≡ (list char)
+                        (tuple <intertype>*) ≡ (record ("𝒊" <intertype>)*) for 𝒊=0,1,...
+                                        bool ≡ (variant (case "true") (case "false"))
```

Comment:
Could we swap these variants? It aligns the implicit numbering with the intuitive numeration of the booleans.

Reply:
Ah, right, yes, good point.